### ISSUE TITLE
[Question] How can the scrapping service can be optimized?

### ISSUE THREAD

**Issue by longnd** [ThinhMaiQuoc/NimbleChallengeFE#6]
## Issue
Good decision was made to handle the scraping service asynchrously by enqueuing each keyword as a background job. However, the scraping service uses Puppeeter to perform the search. The implementation isn't optimal due to the following reasons:
- Puppeteer requires running a Chromium instance which is resource intensive. It is also slow as the full page rendering is needed. To speed up the scraping process, multiple workers can run to consume the jobs. That would make the resource consumption even worst. So, it is not scalable. 
- To workaround Google rate limiting, the scraping service force Puppeeter to wait for random delay, it also increase the execution time.

If we need to improve the implementation, would you suggest a better solution?

**Comment by ThinhMaiQuoc** [comment]
Hi Long,

Thank you so much for your feedback on the scraping service and for highlighting the limitations of the current solution. I've given it some thought and here are a few options for improving our implementation:

1. Switching to Playwright: This tool is more resource-efficient than Puppeteer and better at handling dynamic content. Its advanced features like auto-wait and network activity monitoring could reduce execution time and improve scraping efficiency.

2. Cheerio with Request: For static content, Cheerio could be a lightweight alternative. It's less resource-intensive and faster for simple scraping tasks, although it lacks the ability to handle JavaScript-rendered content like Google's search results.

3. Selenium with Headless Browsers: Selenium offers robust browser automation capabilities and can also handle dynamic content well. It's a mature tool with extensive community support and can be run in headless mode for efficiency.

4. Google's API: Although the project guidelines discourage the use of third-party APIs, Google's own API could be considered as a fallback or supplementary method, especially to cross-verify the scraped data or to handle specific queries that are more challenging to scrape.

In conclusion, while each option has its pros and cons, Playwright stands out as a superior choice for its efficient handling of dynamic content and scalability. It addresses Puppeteer's resource intensity issues through its lighter browser contexts and offers advanced strategies to dynamically navigate Google's rate limits, reducing execution time. Adopting Playwright, complemented by a dynamic rate limiting approach, promises significant improvements in our scraping service's performance and efficiency.

I've created a pull request with the Playwright implementation for the BE repository, and I'm pleased to report that it has significantly enhanced the performance of our scraping service.
https://github.com/ThinhMaiQuoc/NimbleChallengeBE/pull/7

**Comment by longnd** [comment]
Thank you for sharing the ideas for improvement. However, IMO 
1 & 3: (Playwright & Selenium) aren't much different compared to the current solution (Puppeteer) as all of them require using a browser to perform the request (and simulate the user's actions)

4. Google's API: you are right that delegating the work to another party is also a good option, but as mentioned in the code challenge's requirements, we don't expect the use of Google API as the scraping part is most tricky requirement that requires the candidate to showcase their creative & design architecture skills. Also, AFAIK, none of the Google search API return adwords result.

2. Cheerio with Request: it isn't clear to me what you mean by Cheerio with Request. 
but I assume you mean sending an HTTP request directly uisng a HTTP client, e.g. axios and then using Cheerio to parse the content? if it is the case, I belive this is the most lightweight & scalable solution. 
Your concern
> it lacks the ability to handle JavaScript-rendered content like Google's search results.

I think isn't a blocker in this case. Google search result is NOT JS-rendered (client side rendered), it's server-side rendered so it should work. You can test by sending a simple CURL command.

```
curl -A "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3" https://www.google.com/search\?q\=golang
```

as seen in the above CURL command, a simple way to workaround Google rate limiting could be rotating the user agent. Another step could be having a delayed between the jobs, bullJS supports that
https://optimalbits.github.io/bull/#delayed

note that using a delayed job is cleaner than forcing the scraping service to wait as in the current implementation.

**Comment by ThinhMaiQuoc** [comment]
Hi Long, 

Thank you so much for your responses and recommendations for improvement.

Your comment on the Cheerio solution is exactly what I meant and I think the optimal way here is to implement Cheerio with User Agent Rotation.

Regarding using BullJS delayed between jobs, I'm aware of this method, but I actually have several concerns to not use it in my implementation. I agree it'd offer a more efficient and controlled way to manage the scraping tasks. However, managing delayed jobs adds a layer of complexity, if the rate of adding jobs surpasses the processing rate, it might lead to a build-up of pending jobs, potentially causing memory issues and delays. To be honest, those are just my thoughts as I haven't implemented this method ever in practice.

Once again, thanks for your constructive feedback and I'd try your suggestions later to have some hands-on.
